### ISSUE TITLE
Interpreting the input and output of GRU layer

### ISSUE THREAD

**Issue by georgemavrakis-wings** [deephealthproject/pyeddl#57]
Good afternoon,

I have a question concerning the input and output tensors that are returned from GRU layer. More specifically:

I have a data training tensor (X_train) of shape: (10, 7, 24) and a label training tensor (y_train) of shape: (10, 2), which I reshaped into (10, 1, 2) according to nlp_sentiment_gru example. I have constructed the following network and fit the data into it:

```python
in_ = eddl.Input([X_train.shape[2]])    # only the number of features
layer = in_
layer1 = eddl.GRU(layer, units=64)
layer2 = eddl.Constant(l=eddl.Dense(parent=layer1, ndim=2), v=0.03)
out = eddl.Activation(parent=layer2, activation='softmax')

net = eddl.Model(in_=[in_], out=[out])

eddl.build(net, eddl.adam(0.001), ["soft_cross_entropy"], ["categorical_accuracy"], eddl.CS_CPU())

eddl.fit(net, [X_train], [y_train], batch=y_train.shape[0], epochs=1)
```
When I attempt to take the input and output of the network's layers, I notice that the input passed to the input layer has shape: (1,24) and has random values: 

```python
[[ 9.4979256e-03,  4.5805644e-41,  1.2559013e+00,  4.5805644e-41,
  -9.0784385e+20,  8.5954899e+17,  9.4967932e-03,  4.5805644e-41,
   1.2559185e+00,  4.5805644e-41,  3.7693972e-06, -3.2770483e+31,
   9.4980448e-03,  4.5805644e-41,  1.2559357e+00,  4.5805644e-41,
   2.7658451e-24,  5.0690244e+07,  9.4957203e-03,  4.5805644e-41,
   1.2559528e+00,  4.5805644e-41, -2.9805899e-38,  1.9433807e-25]]
```
Therefore, all layers after that, process this vector and the output of the network has shape: (1, 2) and not (10, 2).

I cannot understand what I do wrong.

In addition, when I replace the fit method with train_batch, i.e.:

```python
eddl.train_batch(net, [X_train], [y_train], indices=np.linspace(0, 9, 10, dtype=int).tolist())
```
I get an Incmpatible shape (Tensor::select) traceback.

Thanks in advance for your help!

**Comment by chavicoski** [comment]
Hi. Can you send the complete code that you used to get that input/output tensors of the layers?

**Comment by georgemavrakis-wings** [comment]
> Hi. Can you send the complete code that you used to get that input/output tensors of the layers?

Good morning,

the code I used after the fit of the network, is basically the folllowing:

```python
l = net.layers[1]      # GRU layer
GRU_input = l.input.getdata()
GRU_output = eddl.getOutput(l).getdata()   # to numpy array
```
The above code I used to take the input/output of any layer of the network.

**Comment by chavicoski** [comment]
Hello,

Those tensor shapes are different from what you expect because the EDDL doesn't implement the recurrent nets as other libraries like Tensorflow do. 

The EDDL allways does the unrolling of the net in memory, so when you call the fit function, what is really happening is that the input batches are splited by the sequence dimension (from (10, 7, 24) to 7 tensors of shape (10, 24)) and the EDDL inserts each batch in their corresponding layer timestep. So the input tensor that you get when you call "l.input" is a tensor of shape (batch, in_dim). 

And why you get (1, 24) instead of (10, 24)? Because you are getting the GRU layer from the "net" object wich is not the unrolled net. So the batch size of the input tensor of the GRU in "net" is allways 1 because that net object is never resized to 10, because that resize is made to the unrolled net. Currently we don't have the functionality to get the input/output sequence of an intermediate recurrent layer.

Regarding the random values, as I said previously, that "net" object is not the net that is doing the forward operation, so those values are the original values of the input tensor after its initialization.

Finally, the error in train_batch() is produced because the unrolling is made in the fit function (also in evaluate and predict functions). So the net is not prepared to do the train_batch() because "net" isn't an unrolled net.

Would you like us to implement the functionality to get those inputs/outputs, or did you just have that question about how it works?

**Comment by georgemavrakis-wings** [comment]
Hello,

@chavicoski 

So, the reason why my code does not work is that the functionality to get the input/output operations is not currently implemented for the recurrent layers, because for other layers (Conv1D, Dense etc) it works. Correct?

I am not sure I understand your explanation about the train_batch() function. However, I assume that it has to do with the recurrent layers again, as for other networks it works.

As for your question about the implementation, it would be nice to have it in order to compare the inputs/outputs of the GRU layers with the PyTorch ones.

**Comment by chavicoski** [comment]
Yes, for non-recurrent layers you can use getOutput() to get the real output values.

About the train_batch(), what i was trying to say is that the train_batch() doesn't perform the unrolling so it raises an error because the input shape (10, 7, 24) is not valid.

-"the expected output of the unidirectional GRU layer is that of the last timestep, right?"
-I don't know it for sure. I will confirm you when I know it.

Well, we take note and we will implement that functionality asap. Thank you!

**Comment by RParedesPalacios** [comment]
We have solved it for the recurrent layers

you have to use also getInput to get the input tensor of a layer:

l = net.layers[1]      # GRU layer
GRU_input = eddl.getInput(l).getdata()
GRU_output = eddl.getOutput(l).getdata()   # to numpy array


I guess that Simone has to add the binding for this new functionality

Just in case you are hurry you can getOutput of the previous layer  (the in_)

**Comment by RParedesPalacios** [comment]
Please take into account that if you don't want to use fit for training you can use atomic functions, see for instance example 10 in MNIST:

or(int i=0;i<epochs;i++) {
        printf("Epoch %d\n",i+1);
        reset_loss(net);
        for(int j=0;j<num_batches;j++) {
            // get a batch
            next_batch({x_train,y_train},{x_train_batch,y_train_batch});

            x_train_batch->reshape_({batch_size,28,28}); // time x dim
            y_train_batch->reshape_({batch_size,1,10});

            zeroGrads(net);
            forward(net,{x_train_batch});
            backward(net,{y_train_batch});
            update(net);


            print_loss(net,j);
            printf("\r");

            x_train_batch->reshape_({batch_size,784});

        }
        printf("\n");
    }


https://github.com/deephealthproject/eddl/blob/master/examples/nn/1_mnist/10_mnist_rnn_func.cpp

**Comment by simleo** [comment]
Thanks Álvaro and Roberto!

**Comment by georgemavrakis-wings** [comment]
Good morning,

thank you very much for your effort.

**Comment by RParedesPalacios** [comment]
BTW, train_batch is also ready to deal with recurrent

**Comment by simleo** [comment]
Bindings for `getInput` are in the recently released pyeddl 0.14.0

**Comment by georgemavrakis-wings** [comment]
Hello, thank you.
